utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/BestEffortJsonSchemaEncoder.scala

```diff
@@ -0,0 +1,92 @@
+      }
+      .toList.sequence.map { values => Json.fromFields(values) }
+  }
+
+  private def notNullOrRequired(fieldName: String, value: Any, parentSchema: ObjectSchema): Boolean = {
+    value != null || parentSchema.getRequiredProperties.contains(fieldName)
+  }
+
+  private def encodeCollection(collection: Traversable[_], schema: ArraySchema): EncodeOutput = {
+    collection.map(el => encode(el, schema.getAllItemSchema)).toList.sequence.map(l => Json.fromValues(l))
+  }
+
+  def encodeBasedOnSchema(input: EncodeInput): EncodeOutput = {
+    val (value, schema, fieldName) = input
+    (schema, value) match {
+      case (schema: ObjectSchema, map: scala.collection.Map[String@unchecked, _]) => encodeObject(map.toMap, schema)
+      case (schema: ObjectSchema, map: java.util.Map[String@unchecked, _]) => encodeObject(map.toMap, schema)
+      case (schema: ArraySchema, value: Traversable[_]) => encodeCollection(value, schema)
+      case (schema: ArraySchema, value: java.util.Collection[_]) => encodeCollection(value.toArray, schema)
+      case (schema: StringSchema, value: Any) => encodeStringSchema(schema, value, fieldName)
+      case (_: NumberSchema, value: Long) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Double) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Float) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Int) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: java.math.BigDecimal) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: java.math.BigInteger) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Number) => Valid(jsonEncoder.encode(value.doubleValue()))
+      case (_: NullSchema, null) => Valid(Json.Null)
+      case (_: NullSchema, None) => Valid(Json.Null)
+      case (_: BooleanSchema, value: Boolean) => Valid(Json.fromBoolean(value))
+      case (_: EnumSchema, value: Enum[_]) => Valid(Json.fromString(value.toString))
+      case (null, value: Any) if validationMode == ValidationMode.lax => Valid(jsonEncoder.encode(value))
+      case (_, null) if validationMode != ValidationMode.lax => error(s"Not expected null for field: $fieldName with schema: $schema")
+      case (null, _) if validationMode != ValidationMode.lax => error(s"Not expected null for field: $fieldName with schema: $schema")
+      case (_, _) if validationMode != ValidationMode.lax => error(s"Not expected type: ${value.getClass.getName} for field: $fieldName with schema: $schema")
+    }
+  }
+
+  private def encodeStringSchema(schema: StringSchema, value: Any, fieldName: Option[String] = None) = {
+    (schema.getFormatValidator.formatName(), value) match {
+      case ("date-time", zdt: ZonedDateTime) => Valid(jsonEncoder.encode(zdt))
+      case ("date-time", _: Any) => error(s"Not expected type: ${value.getClass.getName} for field: $fieldName with schema: $schema")
```

Comment:
We could accept more types that can be formatted in the way acceptable by "date-time" format - see BestEffortJsonEncoder and BestEffortAvroEncoder. Smb can create other date type using our helpers. Passing input to output is only one example of usage. We allow all possible conversions between date types. Under some circumstances it will be required to operate on e.g. OffsetDateTime

Reply:
I've added offsetDateTime